src/Gui/ShortcutManager.cpp

```diff
@@ -0,0 +1,455 @@
+                // Not handled. Clear any existing pending actions.
+                timer.stop();
+                for (const auto &info : pendingActions) {
+                    if (info.action)
+                        info.action->setEnabled(true);
+                }
+                pendingActions.clear();
+                lastFocus = nullptr;
+            }
+        }
+        break;
+    case QEvent::ActionChanged:
+        if (auto action = qobject_cast<QAction*>(o)) {
+            auto &index = actionMap.get<0>();
+            auto it = index.find(reinterpret_cast<intptr_t>(action));
+            if (action->shortcut().isEmpty()) {
+                if (it != index.end()) {
+                    QKeySequence oldShortcut = it->key.shortcut;
+                    index.erase(it);
+                    actionShortcutChanged(action, oldShortcut);
+                }
+                break;
+            }
+
+            QByteArray name;
+            if (auto fcAction = qobject_cast<Action*>(action->parent())) {
+                if (fcAction->command() && fcAction->command()->getName())
+                    name = fcAction->command()->getName();
+            }
+            if (name.isEmpty()) {
+                name = action->objectName().size() ?
+                    action->objectName().toUtf8() : action->text().toUtf8();
+                if (name.isEmpty())
+                    name = "~";
+                else
+                    name = QByteArray("~ ") + name;
+            }
+            if (it != index.end()) {
+                if (it->key.shortcut == action->shortcut() && it->key.name == name)
+                    break;
+                QKeySequence oldShortcut = it->key.shortcut;
+                index.replace(it, {action, name});
+                actionShortcutChanged(action, oldShortcut);
+            } else {
+                index.insert({action, name});
+                actionShortcutChanged(action, QKeySequence());
+            }
+        }
+        break;
+    default:
+        break;
+    }
+    return false;
+}
+
+std::vector<std::pair<QByteArray, QAction*>> ShortcutManager::getActionsByShortcut(const QKeySequence &shortcut)
+{
+    const auto &index = actionMap.get<1>();
+    std::vector<std::pair<QByteArray, QAction*>> res;
+    std::multimap<int, const ActionData*, std::greater<int>> map;
+    for (auto it = index.lower_bound(ActionKey(shortcut)); it != index.end(); ++it) {
+        if (it->key.shortcut != shortcut)
+            break;
+        if (it->key.name != "~" && it->action)
+            map.emplace(getPriority(it->key.name), &(*it));
+    }
+    for (const auto &v : map)
+        res.emplace_back(v.second->key.name, v.second->action);
+    return res;
+}
+
+void ShortcutManager::setPriorities(const std::vector<QByteArray> &actions)
+{
+    if (actions.empty())
+        return;
+    // Keep the same top priority of the given action, and adjust the rest. Can
+    // go negative if necessary
+    int current = 0;
```

Comment:
If negative priorities are allowed, does it make sense to initialize `current` to zero?

Reply:
I suppose it can work. It's just that the priority needs to rise from some base level. If the base is the lowest, then we'd have bunch of huge negative integers in the parameters, which would be confusing.

The reason the priority is allow to go negative is because I try to preserve the top priority (among actions with the same shortcuts) while reordering. This is a partial solution to a corner case where the user may have set priority in the past for some action in a workbench that hasn't been loaded yet. Keep the top priority unchanged can provide some degree of stability in this case.